### PR TITLE
Feature/Booked_Donations_Report

### PR DESCRIPTION
**Pull request by evbambly**

### CODE REVIEW

File: packages/common/src/types.ts [hunk 1]
```diff
@@ -131,3 +131,13 @@ export enum BookingChange {
   BOOKED = "BOOKED",
   CANCELLED = "CANCELLED",
 }
+
+export type BookedDonationWithDonor = {
```
Comment (yaron1m) [inline packages/common/src/types.ts:135]: BookedDonationWithDonorDetails

(We don't actually have the donor object here)

File: packages/functions/src/reports/BookedDonationsReport.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+} from "@zm-blood-components/common";
+import { validateAppointmentEditPermissions } from "../coordinator/UserValidator";
+import { getAppointmentsByHospital } from "../dal/AppointmentDataAccessLayer";
+import { getDonors } from "../dal/DonorDataAccessLayer";
+
+export default async function (
+  request: FunctionsApi.GetBookedDonationsInHospitalRequest,
+  callerId: string
+): Promise<FunctionsApi.GetBookedDonationsInHospitalResponse> {
+  await validateAppointmentEditPermissions(
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.ts:15]: This is not enough.
See in files like `GetDonorsHandler` how we have switch-case on the admin role and decide which roles can run this request. 

I think you should also ask Adi who has permissions for this API

File: packages/functions/src/reports/BookedDonationsReport.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  const donorsInAppointments = await getDonors(
+    appointments.map((a) => a.donorId)
+  );
+
+  let bookedDonationsWithDonor: BookedDonationWithDonor[] = [];
+  appointments.forEach((appointment) => {
+    const donor = donorsInAppointments.find(
+      (d) => d.id === appointment.donorId
+    );
+    if (donor) {
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.ts:36]: What if not? could it be?

File: packages/functions/src/reports/BookedDonationsReport.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      const donationWithDonor = {
+        donationId: appointment.id || "",
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.ts:38]: `donationId: appointment.id!`

File: packages/functions/src/reports/BookedDonationsReport.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    appointments.map((a) => a.donorId)
+  );
+
+  let bookedDonationsWithDonor: BookedDonationWithDonor[] = [];
+  appointments.forEach((appointment) => {
+    const donor = donorsInAppointments.find(
+      (d) => d.id === appointment.donorId
+    );
+    if (donor) {
+      const donationWithDonor = {
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.ts:37]: I suggest you change this to:

`const donationWithDonor : BookedDonationWithDonor = {`

File: packages/common/src/functions-api.ts [hunk 6]
```diff
@@ -131,3 +132,13 @@ export interface GetDonorsRequest {}
...
 }
+export declare const GetBookedDonationsInHospitalFunctionName =
+  "getBookedDonationsInHospital";
+export interface GetBookedDonationsInHospitalRequest {
+  hospital: Hospital;
+  fromDateMillis: number;
+  toDateMillis: number;
+}
+export interface GetBookedDonationsInHospitalResponse {
+  donations: BookedDonationWithDonor[];
```
Comment (yaron1m) [inline packages/common/src/functions-api.ts:143]: donationsWithDonorDetails

File: packages/functions/src/reports/BookedDonationsReport.test.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  Functions[FunctionsApi.GetBookedDonationsInHospitalFunctionName]
+);
+
+const COORDINATOR_ID = "GetCoordinatorAppointmentsTestUser";
+const DONOR_ID_1 = "GetCoordinatorAppointmentsTestDonorUser1";
+
+const FUTURE_BOOKED = "GetCoordinatorAppointments_FutureBooked";
+const FUTURE_OTHER_HOSPITAL = "GetCoordinatorAppointments_FutureOtherHospital";
+const FUTURE_NOT_BOOKED = "GetCoordinatorAppointments_FutureNotBooked";
+const FUTURE_DONATION_TOO_FAR = "GetCoordinatorAppointments_FutureTooFar";
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.test.ts:31]: Change the names of all the users and appointments to match the function name

File: packages/functions/src/reports/BookedDonationsReport.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  BookedDonationWithDonor,
+  DbAppointment,
+  FunctionsApi,
+  Hospital,
+} from "@zm-blood-components/common";
+import { validateAppointmentEditPermissions } from "../coordinator/UserValidator";
+import { getAppointmentsByHospital } from "../dal/AppointmentDataAccessLayer";
+import { getDonors } from "../dal/DonorDataAccessLayer";
+
+export default async function (
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.ts:11]: File name should be BookedDonationInHospitalReportHandler

File: packages/functions/src/reports/BookedDonationsReport.test.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    getDate(16),
+    Hospital.TEL_HASHOMER,
+    DONOR_ID_1
+  );
+
+  const res = await callFunction(COORDINATOR_ID);
+
+  let appointments = res.donations;
+  expect(appointments).toHaveLength(1);
+  expect(appointments[1].donationId).toEqual(FUTURE_NOT_BOOKED);
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.test.ts:100]: 1. Did you run these tests locally? Did they pass? Cause I see OutOfBoundException here (should be [0] not [1]).
2. Also check the other fields in the response

File: packages/functions/src/reports/BookedDonationsReport.test.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    FUTURE_NOT_BOOKED,
+    getDate(2),
+    Hospital.TEL_HASHOMER,
+    ""
+  );
+  await saveAppointment(
+    FUTURE_OTHER_HOSPITAL,
+    getDate(2),
+    Hospital.ASAF_HAROFE
+  );
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationsReport.test.ts:88]: Should you add donor id?

File: package.json [hunk 11]
```diff
@@ -39,7 +39,7 @@
     "classnames": "^2.2.6",
     "date-fns": "^2.0.0-beta.5",
     "dayjs": "^1.10.4",
-    "firebase": "^8.2.4",
+    "firebase": "^8.5.0",
```
[the remark this answers is not in the document]
Comment (evbambly) [reply package.json:42]: The != operator for a where() query was introduced in v8.5

File: packages/functions/src/reports/BookedDonationsReport.ts (hunk 2 above)
Comment (evbambly) [reply packages/functions/src/reports/BookedDonationsReport.ts]: Currently undefined, 
ZM coordinators are currently the only relevant group.
Marked //TODO

File: packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  request: FunctionsApi.GetBookedDonationsInHospitalRequest,
+  callerId: string
+): Promise<FunctionsApi.GetBookedDonationsInHospitalResponse> {
+  const coordinator = await getCoordinator(callerId);
+  if (!coordinator) {
+    console.error("Could not find calling user", callerId);
+    throw Error(`User ${callerId} is not an admin`);
+  }
+
+  // TODO switch (coordinator.role) && by requested hospital
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts:20]: Implement?

File: packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import { getAppointmentsByHospital } from "../dal/AppointmentDataAccessLayer";
+import { getDonors } from "../dal/DonorDataAccessLayer";
+
+export default async function (
+  request: FunctionsApi.GetBookedDonationsInHospitalRequest,
+  callerId: string
+): Promise<FunctionsApi.GetBookedDonationsInHospitalResponse> {
+  const coordinator = await getCoordinator(callerId);
+  if (!coordinator) {
+    console.error("Could not find calling user", callerId);
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts:16]: No test is testing this condition

File: packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts [hunk 14]
```diff
@@ -17,7 +17,7 @@ export default async function (
     throw Error(`User ${callerId} is not an admin`);
   }
 
-  // TODO switch (coordinator.role) && by requested hospital
+  // TODO implement switch (coordinator.role) && by requested hospital
```
Comment (yaron1m) [inline packages/functions/src/reports/BookedDonationInHospitalReportHandler.ts:20]: Haha I meant you need to implement this :P

Why did you chose to add a TODO and not write the code?

**Review by yaron1m — APPROVED** [review]

**Pull request merged**